### ISSUE TITLE
Reduce memory retained by parsed Typography typefaces

### ISSUE THREAD

**Issue by charlesroddie** [verybadcat/CSharpMath#107]
CSharpMath retains parsed Typography Typeface data for bundled and custom fonts. The original profile showed retained memory far above the source OTF sizes, but it predates the current dependency snapshot; establish a current baseline before changing the loading strategy.

Acceptance criteria:

- Add a reproducible benchmark/profile for startup allocations and retained memory after rendering representative math and text.
- Identify which OpenType tables, glyph instructions, outlines, and caches remain reachable after loading.
- Lazily load or release data that the renderer does not need after extracting metrics/outlines.
- Verify repeated rendering does not reparse fonts or grow retained caches without bound.
- Keep custom LocalTypefaces and existing rendering output working.
- Record before/after measurements for the bundled fonts and at least one custom font.

Precomputing the bundled font data so it need not be parsed at runtime is the separate approach tracked in #191.

**Comment by Happypig375** [comment]
Currently CSharpMath stores Typeface objects in memory which is basically an entire otf file parsed into table objects. latinmodern-math.otf weighs 717KiB, and AMS-Capital-Blackboard-Bold.otf weighs 8.51KiB. However, the inclusive size shown is an order of magnitude larger than the size of the source files.

**Comment by charlesroddie** [comment]
Presumably the issue involves:
1. Loading an entire font at a time rather than characters lazily
2. A memory-inefficient representation of a character

Are both of these issues for the typography repo, or is there anything that can be improved here?

**Comment by Happypig375** [comment]
@prepare Thoughts?

**Comment by prepare** [comment]
I think  some objects could be unloaded/dispose
eg. List<Type2Instruction> could be unload after you create a glyph outline from it.

but this I have to test it first

**Comment by prepare** [comment]
please visit and test 'Trimmable features' again, see=> https://github.com/LayoutFarm/Typography/issues/187

**Comment by Happypig375** [comment]
So there are two ways forward with the Typography submodule:
- Drop LayoutFarm/Typography#192 and update
- Merge latest Typography into LayoutFarm/Typography#192 and resolve conflicts (and do them again for future updates)

@prepare @charlesroddie Thoughts?

**Comment by charlesroddie** [comment]
I don't believe CSharpMath is affected much by the NRT PR LayoutFarm/Typography#192. That PR makes it easier for Typography to control NRE bugs and to do further development without having to worry about NREs. However it's not worth forking Typography just for that. If the @prepare prefers to do more work to implement features safely that is up to him!

CSharpMath should depend on Typography master branch if possible, through the submodule for now, and through a nuget package when Typography stabilizes.

Note that it can't use the current state of Typography master as UWP support is broken by the netstandard2.1 change.

**Comment by prepare** [comment]
Hello 
> Note that it can't use the current state of Typography master as UWP support is broken by the netstandard2.1 change.

I move back to netstandard 2.0, please check again :)

https://github.com/LayoutFarm/Typography/commit/0899ddfb4294325bc3f9bc06909bd3bee8323eca

**Comment by Happypig375** [comment]
@charlesroddie How the NRT PR affects CSharpMath can be seen through the difference between using null-oblivious and null-explicit APIs. By dropping the current submodule reference to LayoutFarm/Typography#192, all Typography APIs currently is use will start ignoring accidental nulls.

**Comment by Happypig375** [comment]
@prepare @charlesroddie Seems that LayoutFarm/Typography#192 will never be merged. Should we drop it and update or should we resolve merge conflicts?

**Comment by charlesroddie** [comment]
That is a shame if true. @prepare it would be good to have your thoughts on that PR even if in an initial form. Maintaining a branch or fork of typography for use in CSharpMath would require a lot of work on @Happypig375 's part, and is unfeasible to maintain indefinitely IMO. So if it will really never be merged then dropping and accepting a less safe typography library for use in CSharpMath is inevitable.

**Comment by prepare** [comment]
@Happypig375 👍 

I want to confirm that the https://github.com/LayoutFarm/Typography/pull/192 is NOT your waste.

But I can't merge it directly now because of this https://github.com/LayoutFarm/Typography/pull/192#issuecomment-635820167 too.

---
By the way, 
The Typography+PixelFarm is now under refactoring-code-arrangement phase again,
 
It will be released this week.
